tests/scenario/test_ingress_per_app.py

```diff
@@ -0,0 +1,172 @@
+# GIVEN a charm with ingress impl'd
+# WHEN a relation with traefik is formed
+# THEN traefik's config file's `server` section has all the units listed
+# AND WHEN the charm rescales
+# THEN the traefik config file is updated
+
+import tempfile
+from pathlib import Path
+
+import pytest
+import yaml
+from ops import pebble
+from scenario import Container, Context, Model, Mount, Relation, State
+
+
+@pytest.fixture
+def context(traefik_charm):
+    return Context(charm_type=traefik_charm)
+
+
+@pytest.fixture
+def model():
+    return Model(name="test-model")
+
+
+@pytest.fixture
+def temp_opt():
+    with tempfile.TemporaryDirectory() as td:
+        yield Path(td)
+
+
+@pytest.fixture
+def traefik_container(temp_opt):
+    layer = pebble.Layer(
+        {
+            "summary": "Traefik layer",
+            "description": "Pebble config layer for Traefik",
+            "services": {
+                "traefik": {
+                    "override": "replace",
+                    "summary": "Traefik",
+                    "command": '/bin/sh -c "/usr/bin/traefik | tee /var/log/traefik.log"',
+                    "startup": "enabled",
+                },
+            },
+        }
+    )
+
+    opt = Mount("/opt/", temp_opt)
+
+    return Container(
+        name="traefik",
+        can_connect=True,
+        layers={"traefik": layer},
+        service_status={"traefik": pebble.ServiceStatus.ACTIVE},
+        mounts={"opt": opt},
+    )
+
+
+@pytest.mark.parametrize("port, host", ((80, "1.1.1.1"), (81, "10.1.10.1")))
+@pytest.mark.parametrize("event_name", ("joined", "changed"))
+def test_ingress_per_app_created(
+    context, port, host, model, traefik_container, event_name, temp_opt
+):
+    """Check the config when a new ingress per leader is created or changes (single remote unit)."""
+    ipl = Relation(
+        "ingress",
+        remote_app_data={
+            "model": "test-model",
+            "name": "remote/0",
+            "mode": "http",
+        },
+        remote_units_data={0: {"port": str(port), "host": host}},
+        relation_id=1,
+    )
+    state = State(
+        model=model,
+        config={"routing_mode": "path", "external_hostname": "foo.com"},
+        containers=[traefik_container],
+        relations=[ipl],
+    )
+
+    event = getattr(ipl, f"{event_name}_event")
```

Comment:
It would be very helpful to have some comments inline, for example:
```suggestion
    # WHEN any relevant event fires
    event = getattr(ipl, f"{event_name}_event")
```